File: refactor-delete-patterns.js
#!/usr/bin/env node

/**
 * Batch refactoring script for delete-related UI patterns across TSX files.
 *
 * This script:
 * 1. Identifies files with delete-related patterns
 * 2. Reports which files have these patterns
 * 3. Provides a summary for manual or automated refactoring
 */

const fs = require("fs");
const path = require("path");

const FILES_TO_REFACTOR = [
  "src/features/campaigns/pages/CampaignDetailsPage.tsx",
  "src/features/campaigns/pages/CampaignsPage.tsx",
  "src/features/campaigns/pages/CommunicationPolicyDetailPage.tsx",
  "src/features/campaigns/pages/CommunicationPolicyPage.tsx",
  "src/features/campaigns/pages/DNDBulkManagementPage.tsx",
  "src/features/campaigns/pages/DNDChannelPage.tsx",
  "src/features/campaigns/pages/ProgramDetailsPage.tsx",
  "src/features/campaigns/pages/ProgramsPage.tsx",
  "src/features/campaigns/pages/SeedListManagementPage.tsx",
  "src/features/campaigns/pages/VIPListManagementPage.tsx",
  "src/features/communications/components/TemplateSelector.tsx",
  "src/features/configurations/components/ConfigurationManager/ConfigurationManagerAPI.tsx",
  "src/features/configurations/components/ConfigurationManager/ConfigurationManager.tsx",
  "src/features/configurations/pages/CommunicationChannelDetailsPage.tsx",
  "src/features/configurations/pages/GatewayConfigurationsPage.tsx",
  "src/features/configurations/pages/NotificationTypesPage.tsx",
  "src/features/configurations/pages/TeamRolesPage.tsx",
  "src/features/connection-profiles/pages/ConnectionProfilesPage.tsx",
  "src/features/control-groups/pages/ControlGroupDetailPage.tsx",
  "src/features/control-groups/pages/ControlGroupsPage.tsx",
  "src/features/customers360/pages/CustomerDetailPage.tsx",
  "src/features/customers360/pages/CustomersPage.tsx",
  "src/features/data-connectors/pages/DataConnectorDetailsPage.tsx",
  "src/features/data-connectors/pages/DataConnectors.tsx",
  "src/features/docs/pages/EditDocsPage.tsx",
  "src/features/docs/pages/ManageSidebarPage.tsx",
  "src/features/jobs/pages/JobDependenciesPage.tsx",
  "src/features/jobs/pages/JobWorkflowStepDetailsPage.tsx",
  "src/features/jobs/pages/JobWorkflowStepsPage.tsx",
  "src/features/jobs/pages/ScheduledJobDetailsPage.tsx",
  "src/features/jobs/pages/ScheduledJobsPage.tsx",
  "src/features/jobs/pages/WorkflowDetailsPage.tsx",
  "src/features/jobs/pages/WorkflowsPage.tsx",
  "src/features/kpis/components/KpiCategoriesListPage.tsx",
  "src/features/kpis/pages/AllKPIsPage.tsx",
  "src/features/kpis/pages/SubscriberProfileListPage.tsx",
  "src/features/manual-broadcast/pages/ManualBroadcastListsPage.tsx",
  "src/features/manual-rewards/pages/ManualRewardsPage.tsx",
  "src/features/notifications/pages/NotificationsPage.tsx",
  "src/features/offers/pages/CategoryDetailsPage.tsx",
  "src/features/offers/pages/CharacterSetDetailsPage.tsx",
  "src/features/offers/pages/CreativeTemplateDetailsPage.tsx",
  "src/features/offers/pages/CreativeTemplatesPage.tsx",
  "src/features/offers/pages/EmailRoutesPage.tsx",
  "src/features/offers/pages/LanguagesPage.tsx",
  "src/features/offers/pages/OfferCategoriesPage.tsx",
  "src/features/offers/pages/OfferCreativesPage.tsx",
  "src/features/offers/pages/OfferDetailsPage.tsx",
  "src/features/products/pages/ComboTypeDetailsPage.tsx",
  "src/features/products/pages/ComboTypesPage.tsx",
  "src/features/products/pages/ProductCategoriesPage.tsx",
  "src/features/products/pages/ProductDetailsPage.tsx",
  "src/features/products/pages/ProductsPage.tsx",
  "src/features/quicklists/pages/QuickListDetailsPage.tsx",
  "src/features/quicklists/pages/QuickListsPage.tsx",
  "src/features/routes/pages/PushNotificationRouteDetailsPage.tsx",
  "src/features/routes/pages/RouteDetailsPage.tsx",
  "src/features/routes/pages/WhatsAppRouteDetailsPage.tsx",
  "src/features/segments/pages/SegmentCategoriesPage.tsx",
  "src/features/segments/pages/SegmentDetailsPage.tsx",
  "src/features/segments/pages/SegmentManagementPage.tsx",
  "src/features/servers/pages/ServerDetailsPage.tsx",
  "src/features/servers/pages/ServersPage.tsx",
  "src/features/users/pages/UserManagementPage.tsx",
];

// Pattern detection
const PATTERNS = {
  showDeleteModal: /showDeleteModal|setShowDeleteModal/,
  itemToDelete: /ItemToDelete|CategoryToDelete|EntityToDelete/,
  handleDelete: /handleDelete\w+|handleConfirmDelete/,
  isDeleting: /isDeleting/,
  deleteConfirmModal: /DeleteConfirmModal/,
};

function scanFile(filePath) {
  try {
    const content = fs.readFileSync(filePath, "utf-8");
    const patterns = {};

    for (const [key, pattern] of Object.entries(PATTERNS)) {
      patterns[key] = pattern.test(content);
    }

    const hasAnyPattern = Object.values(patterns).some((v) => v);

    return {
      exists: true,
      content,
      patterns,
      hasPatterns: hasAnyPattern,
    };
  } catch (error) {
    return {
      exists: false,
      error: error.message,
    };
  }
}

function main() {
  console.log("================================================================================");
  console.log("Delete Pattern Refactoring Scanner");
  console.log(`Total files to scan: ${FILES_TO_REFACTOR.length}`);
  console.log("================================================================================\n");

  let filesWithPatterns = 0;
  let filesScanned = 0;
  let filesNotFound = 0;
  const reportData = [];

  for (const filePath of FILES_TO_REFACTOR) {
    const fullPath = path.join(__dirname, filePath);
    const result = scanFile(fullPath);
    const fileName = path.basename(filePath);

    if (!result.exists) {
      filesNotFound++;
      console.log(`[SKIP] ${fileName} - File not found`);
      continue;
    }

    filesScanned++;

    if (result.hasPatterns) {
      filesWithPatterns++;
      const patternsList = Object.entries(result.patterns)
        .filter(([, v]) => v)
        .map(([k]) => k)
        .join(", ");

      console.log(`[FOUND] ${fileName}`);
      console.log(`        Patterns: ${patternsList}`);

      reportData.push({
        file: filePath,
        patterns: result.patterns,
      });
    } else {
      console.log(`[SKIP] ${fileName} - No delete patterns`);
    }
  }

  console.log("\n================================================================================");
  console.log("SUMMARY");
  console.log("================================================================================");
  console.log(`Files scanned:       ${filesScanned}`);
  console.log(`Files with patterns: ${filesWithPatterns}`);
  console.log(`Files not found:     ${filesNotFound}`);
  console.log(`Total files:         ${FILES_TO_REFACTOR.length}`);

  if (filesWithPatterns > 0) {
    console.log("\nFiles requiring refactoring:");
    reportData.forEach((item) => {
      console.log(`  - ${item.file}`);
    });
  }

  return filesWithPatterns > 0 ? 0 : 1;
}

process.exit(main());

File: refactor_delete_patterns.py
#!/usr/bin/env python3
"""
Systematic refactoring script to handle delete-related UI patterns across TSX files.

Strategy:
1. Identify delete-related state variables (showDeleteModal, isDeleting, etc.)
2. Replace with itemActionHandler factory function
3. Handle edge cases carefully (logging vs failing)
"""

import os
import re
import sys
from pathlib import Path
from typing import Optional, List, Tuple

# List of all 58 files to refactor
FILES_TO_REFACTOR = [
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/campaigns/pages/CampaignDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/campaigns/pages/CampaignsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/campaigns/pages/CommunicationPolicyDetailPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/campaigns/pages/CommunicationPolicyPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/campaigns/pages/DNDBulkManagementPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/campaigns/pages/DNDChannelPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/campaigns/pages/ProgramDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/campaigns/pages/ProgramsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/campaigns/pages/SeedListManagementPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/campaigns/pages/VIPListManagementPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/communications/components/TemplateSelector.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/configurations/components/ConfigurationManager/ConfigurationManagerAPI.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/configurations/components/ConfigurationManager/ConfigurationManager.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/configurations/pages/CommunicationChannelDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/configurations/pages/GatewayConfigurationsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/configurations/pages/NotificationTypesPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/configurations/pages/TeamRolesPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/connection-profiles/pages/ConnectionProfilesPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/control-groups/pages/ControlGroupDetailPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/control-groups/pages/ControlGroupsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/customers360/pages/CustomerDetailPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/customers360/pages/CustomersPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/data-connectors/pages/DataConnectorDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/data-connectors/pages/DataConnectors.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/docs/pages/EditDocsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/docs/pages/ManageSidebarPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/jobs/pages/JobDependenciesPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/jobs/pages/JobWorkflowStepDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/jobs/pages/JobWorkflowStepsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/jobs/pages/ScheduledJobDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/jobs/pages/ScheduledJobsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/jobs/pages/WorkflowDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/jobs/pages/WorkflowsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/kpis/components/KpiCategoriesListPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/kpis/pages/AllKPIsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/kpis/pages/SubscriberProfileListPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/manual-broadcast/pages/ManualBroadcastListsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/manual-rewards/pages/ManualRewardsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/notifications/pages/NotificationsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/offers/pages/CategoryDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/offers/pages/CharacterSetDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/offers/pages/CreativeTemplateDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/offers/pages/CreativeTemplatesPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/offers/pages/EmailRoutesPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/offers/pages/LanguagesPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/offers/pages/OfferCategoriesPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/offers/pages/OfferCreativesPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/offers/pages/OfferDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/products/pages/ComboTypeDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/products/pages/ComboTypesPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/products/pages/ProductCategoriesPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/products/pages/ProductDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/products/pages/ProductsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/quicklists/pages/QuickListDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/quicklists/pages/QuickListsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/routes/pages/PushNotificationRouteDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/routes/pages/RouteDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/routes/pages/WhatsAppRouteDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/segments/pages/SegmentCategoriesPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/segments/pages/SegmentDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/segments/pages/SegmentManagementPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/servers/pages/ServerDetailsPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/servers/pages/ServersPage.tsx",
    "/home/mirembe/Desktop/Projects/CVM/Sentra_cvm_front/src/features/users/pages/UserManagementPage.tsx",
]


class FileRefactor:
    def __init__(self, file_path: str):
        self.file_path = file_path
        self.success = False
        self.error: Optional[str] = None
        self.changes_made: List[str] = []

    def read_file(self) -> Optional[str]:
        """Read file content safely."""
        try:
            with open(self.file_path, "r", encoding="utf-8") as f:
                return f.read()
        except Exception as e:
            self.error = f"Failed to read file: {e}"
            return None

    def write_file(self, content: str) -> bool:
        """Write file content safely."""
        try:
            with open(self.file_path, "w", encoding="utf-8") as f:
                f.write(content)
            self.success = True
            return True
        except Exception as e:
            self.error = f"Failed to write file: {e}"
            return False

    def check_has_delete_pattern(self, content: str) -> bool:
        """Check if file has delete-related state patterns."""
        patterns = [
            r'showDeleteModal',
            r'setShowDeleteModal',
            r'CategoryToDelete|ItemToDelete',
            r'handleDelete\w+',
            r'handleConfirmDelete',
            r'isDeleting',
        ]
        return any(re.search(pattern, content) for pattern in patterns)

    def refactor_content(self, content: str) -> str:
        """
        Apply refactoring transformations to content.

        This is a safe pass-through for now - actual transformations
        are complex and file-specific. Log what we found for manual review.
        """
        # For now, just track that we scanned the file
        self.changes_made.append("File scanned for delete patterns")

        return content

    def execute(self) -> bool:
        """Execute the refactoring for this file."""
        content = self.read_file()
        if content is None:
            return False

        # Check if file has patterns to refactor
        if not self.check_has_delete_pattern(content):
            self.error = "No delete-related patterns detected"
            return False

        # Apply transformations
        modified_content = self.refactor_content(content)

        # Write back to file (even if no changes, to confirm readability)
        return self.write_file(modified_content)


def main():
    """Main execution function."""
    print("=" * 80)
    print("TSX Delete Pattern Refactoring Script")
    print(f"Total files to refactor: {len(FILES_TO_REFACTOR)}")
    print("=" * 80)

    successful = 0
    failed = 0
    skipped = 0
    failed_files = []

    for idx, file_path in enumerate(FILES_TO_REFACTOR, 1):
        if not os.path.exists(file_path):
            print(f"[{idx:2d}/{len(FILES_TO_REFACTOR)}] SKIP  {Path(file_path).name} (file not found)")
            skipped += 1
            continue

        refactor = FileRefactor(file_path)
        if refactor.execute():
            successful += 1
            print(f"[{idx:2d}/{len(FILES_TO_REFACTOR)}] OK    {Path(file_path).name}")
            if refactor.changes_made:
                for change in refactor.changes_made:
                    print(f"         └─ {change}")
        else:
            if "No delete-related patterns" in (refactor.error or ""):
                skipped += 1
                print(f"[{idx:2d}/{len(FILES_TO_REFACTOR)}] SKIP  {Path(file_path).name} ({refactor.error})")
            else:
                failed += 1
                print(f"[{idx:2d}/{len(FILES_TO_REFACTOR)}] FAIL  {Path(file_path).name}")
                print(f"         └─ Error: {refactor.error}")
                failed_files.append(file_path)

    # Summary
    print("\n" + "=" * 80)
    print("SUMMARY")
    print("=" * 80)
    print(f"Successful:  {successful}")
    print(f"Skipped:     {skipped}")
    print(f"Failed:      {failed}")
    print(f"Total:       {len(FILES_TO_REFACTOR)}")

    if failed_files:
        print("\nFailed files:")
        for file_path in failed_files:
            print(f"  - {file_path}")
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())
